Read-only Refunds, part 1

mindgraffiti opened the pull request:
This PR takes the first steps to displaying refunds UI in the Order Details screen. Please excuse the large PR...I realized half way through my additions that OrderDetailsDataSource was huge and overwhelming, and refactored it.

## Changes
- Discovered more Yosemite work that I previously missed. Added `OrderRefundsCondensed` to Yosemite.
- Originally, I publicly exposed the footerLabel in `TwoColumnHeadlineFooterTableViewCell`. But then I realized it violates the open / closed principle.
- Created a new tableviewcell, `TwoColumnHeadlineAttributedFooterTableViewCell`, to remedy this
- ~Refactored `OrderDetailsDataSource`. Broke it into 4 files: main, Constants, Cells, and Sections. If there are parts that need moved (so it makes sense), please let me know!~
- Feature flagged Refunds. 
- The Refund cell(s) appear on Order Details for an order containing 1 or more refunds.
- The Net cell appears on Order Details when a refund is present. This cell has not been fully configured, all it has is "Net" for the title.

*Known Issues:*
- Currency formatting for Refund & Net cell full configuration to come in another PR
- Net footnote text to be removed in another PR
- View details > Refund Details screen to come in another PR

## To test
1. Find an order in your store
2. Create a partial refund on the desktop web interface
3. Navigate to the correct order number
4. Expected result: Refund cell(s) display. The Net amount cell appears (but doesn't have any real info yet).

iPhone 8:
<kbd>
<img src="https://user-images.githubusercontent.com/1062444/69279353-c7f01500-0ba9-11ea-9470-0d24ecae569f.png" width="350" />
</kbd>

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mindgraffiti committed:
Delete commented out code

mindgraffiti committed:
Add the typealias for Yosemite to Storage: OrderCondensedRefund and Refund

mindgraffiti committed:
Begin syncing refunds in the OrderDetailsVC

mindgraffiti committed:
Add the syncRefunds method to the viewModel

mindgraffiti committed:
Add the refundResultsController to Order Detail's ResultsControllers class.

mindgraffiti committed:
Add the refund and netAmount cases to the enum

mindgraffiti committed:
Add the enums to the cell configuration method

mindgraffiti committed:
Outline cell configuration methods for refund and net amount

mindgraffiti committed:
Add some configurations to configureRefund()

mindgraffiti committed:
Add refunds to payments section

mindgraffiti committed:
Add lookUpRefund() method

mindgraffiti committed:
Extend the TwoColumnHeadlineFootnote cell to accept attributedText

mindgraffiti committed:
Add refunds to the viewModel

mindgraffiti committed:
Add more to configureRefund

mindgraffiti committed:
Add refund and refund summary outline to payment details view model

mindgraffiti committed:
Add to DateHelpers - create a medium length, localized date formatter

mindgraffiti committed:
Remove unused variable

mindgraffiti committed:
Correct errors

mindgraffiti committed:
Update project scheme files to remove warning

mindgraffiti committed:
Merge branch 'develop' into issue/1099-read-only-refunds

mindgraffiti committed:
Alphabetize

mindgraffiti committed:
Register RefundStore in Services

mindgraffiti committed:
Add the refunds field to the order details request

mindgraffiti committed:
Update self check

mindgraffiti committed:
Add OrderRefundsCondensed to Yosemite

mindgraffiti committed:
Add a private property to access order.orderRefundsCondensed

mindgraffiti committed:
First attempt at localizing and styling an attributed string

mindgraffiti committed:
Merge branch 'develop' into issue/1099-read-only-refunds

mindgraffiti committed:
Get the Refund details text to show up on the cell

mindgraffiti committed:
Revert changes to the TwoColumnHeadlineFootnote cell

mindgraffiti committed:
Create a new cell: TwoColumnHeadline *Attributed* Footnote cell

Modifying the original TwoColumnHeadlineFootnote cell violates the Open / Closed principle.

mindgraffiti committed:
Use the new cell for Refunds

mindgraffiti committed:
There is no need for default styling of the attributed string

Instead, style the link piece and leave the rest to be styled by the default cell settings.

mindgraffiti committed:
Set member variables to the default (internal) file access 

instead of private. This prepares the file so we can break it down into reasonable chunks.

mindgraffiti committed:
Register the new TwoColumnHeadlineAttributedFootnote cell

mindgraffiti committed:
Refactor so Constants are in their own file

mindgraffiti committed:
Refactor Cell data to be in its own file

mindgraffiti committed:
Refactor Section data to be in its own file

mindgraffiti committed:
Feature flag Refunds

mindgraffiti committed:
Merge branch 'develop' into issue/1099-read-only-refunds

mindgraffiti committed:
Code styling clean up

mindgraffiti committed:
Found extra whitespace

mindgraffiti committed:
Code clean up continued

jaclync commented on Yosemite/Yosemite/Stores/OrderStore.swift:
```diff
@@ -281,6 +281,7 @@ private extension OrderStore {
             handleOrderItems(readOnlyOrder, storageOrder, storage)
             handleOrderCoupons(readOnlyOrder, storageOrder, storage)
             handleOrderShippingLines(readOnlyOrder, storageOrder, storage)
+            handleOrderRefundsCondensed(readOnlyOrder, storageOrder, storage)
```
suggestion: could add a unit test to verify this change

mindgraffiti replied:
Will do!

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource.swift:
```diff
@@ -27,18 +34,23 @@ final class OrderDetailsDataSource: NSObject {
     }
 
-    private var summaryDateCreated: String {
+    /// For example, Oct 1, 2019 at 2:31 PM
```
I like the examples!

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource+Sections.swift:
```diff
@@ -0,0 +1,234 @@
+        let payment: Section = {
+            var rows: [Row] = [.payment, .customerPaid]
+
+            if FeatureFlag.refunds.enabled {
```
I believe we started using the feature flag service from the `ServiceLocator` that enables mocking, like `ServiceLocator.featureFlagService.isFeatureFlagEnabled(. refunds)` with the feature flag configured in the default implementation `DefaultFeatureFlagService`

mindgraffiti replied:
Thank you!

jaclync reviewed (commented):
It's really cool to see the refund info!

**Code:**
I'm not so sure about separating core code into extensions with properties or functions that originally private becoming public - was the reason for separating them because of the file size or any other reasons? Due to the moved code, it was a bit hard to tell the diffs from the refund work vs. the original code in `OrderDetailsDataSource+Cells`. In the future, I'd suggest a separate PR for refactoring with the same functionality. I don't have a concrete idea for refactoring the `OrderDetailsDataSource` right now, maybe we can come up with another way to separate the logic into single functionality components (like moving the configuration functions to a view model / cell extension) in the future.

**Testing:**
I didn't find the "net amount" field on wp-admin, it's probably the net payout = original amount - total refund? from my testing, the net amount is the same as the original amount:

<img src="https://user-images.githubusercontent.com/1945542/69041324-49b93480-0a2a-11ea-9d16-3e7ac357fb10.png" width="350" />

mindgraffiti commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource+Cells.swift:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    func configureRefund(cell: TwoColumnHeadlineAttributedFootnoteTableViewCell, at indexPath: IndexPath) {
+//        let row = rowAtIndexPath(indexPath)
```
It will be used in a different method. Will delete for now.

mindgraffiti commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource+Cells.swift:
```diff
@@ -0,0 +1,257 @@
+
+    func configureRefund(cell: TwoColumnHeadlineAttributedFootnoteTableViewCell, at indexPath: IndexPath) {
+//        let row = rowAtIndexPath(indexPath)
+        let condensedRefund = condensedRefunds[indexPath.row - 2] // TODO-thuy: minus two should be constants
```
This needs refactored into `lookUpCondensedRefunds(_ indexPath)`. I had so much code in here already I didn't want to add more. I can convert this now if you prefer. 

An order can have multiple condensed refunds. In order to configure the correct row, I have to look up the correct refund. It's currently subtracting out 2 rows because the refunds are listed below the Payment Summary cell and the Paid by Customer cell are added to this section first.

jaclync replied:
Got it, the TODO sounds good 👍

mindgraffiti commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource+Cells.swift:
```diff
@@ -0,0 +1,257 @@
+
+    func configureNetAmount(cell: TwoColumnHeadlineFootnoteTableViewCell) {
+        cell.leftText = Titles.netAmount
+//        cell.rightText = viewModel.netAmount
```
I should have marked this section as `TODO-thuy: configure the Net amount cell in the next PR.`

mindgraffiti commented on WooCommerce/Classes/ViewRelated/ReusableViews/TwoColumnHeadlineAttributedFootnoteTableViewCell.swift:
```diff
@@ -0,0 +1,77 @@
+        set {
+            footnoteLabel?.attributedText = newValue
+        }
+    }
```
Originally, I publicly exposed the footerLabel in `TwoColumnHeadlineFooterTableViewCell`. But then I realized it violates the single responsibility principle and the open / closed principle, because I would have to open the `footnoteLabel` for public access and create the new `footnoteAttributedText` variable. Now footnoteLabel has 2 responsibilities. Either display text or display attributed text.

Adding a new var `footnoteAttributedText` doesn't work by itself, because I have to check if the `footnoteLabel.text` property is nil, remove the original label styling, and apply a different label style. All of those actions should be done in footnoteAttributedText's `didSet` method, but `didSet` can't be used since there is already a getter.

I created a new tableviewcell, `TwoColumnHeadlineAttributedFooterTableViewCell` as a solution to all of these problems.

jaclync replied:
I think it's pretty common to have multiple ways of styling a view for similar use cases, the name "TwoColumnHeadlineFooterTableViewCell" feels pretty generic and I wouldn't be surprised if there is more than one way to set the headline/footer.

if just adding the same code
```
var footnoteAttributedText: NSAttributedString? {
        get {
            return footnoteLabel?.attributedText
        }
        set {
            footnoteLabel?.attributedText = newValue
        }
}
```

to `TwoColumnHeadlineFootnoteTableViewCell` and using it for the Refund cells, I feel that might work?

Regarding the getters, I checked the Apple doc:

About [`UILabel`'s `text` property](https://developer.apple.com/documentation/uikit/uilabel/1620538-text):
> Assigning a new value to this property also replaces the value of the attributedText property with the same text, although without any inherent style attributes.

About [`UILabel`'s `attributedText ` property](https://developer.apple.com/documentation/uikit/uilabel/1620542-attributedtext):
> Assigning a new value to this property also replaces the value of the text property with the same string data, although without any formatting information. In addition, assigning a new a value updates the values in the font, textColor, and other style-related properties so that they reflect the style information starting at location 0 in the attributed string.

From the docs above, it seems like setting either `text` or `attributedText` will update the label's text with or without styles. So I don't think it'd require the knowledge of the other property. Also, the getters aren't really used and the setters could also be in a function form like `func updateFootnoteText(text: String?)` / `func updateFootnoteAttributedText(footnoteAttributedText: NSAttributedString?)`

My concern for creating another cell with the same layout/style with the exception of a setter is that whenever we update the layout or style in one cell, we'd have to remember to update the other cell and not every dev might be aware of this.

mindgraffiti replied:
I disagree. There is less cost to updating the layout and style in a separate cell than there is with maintaining a single cell that does too many things and will need refactored in the future. @pmusolino would you be a tie-breaker? I believe both @jaclync and I would agree that we will accept whatever decision you make.

mindgraffiti replied:
Edit: I apologize for the misunderstanding. The whole point of the new cell was to **remove** the `.footerText` variable so the cell was not doing two things. I have edited this new cell to demonstrate what I was originally thinking.

pmusolino replied:
To be honest, since the meaning of these two cells is the same, I would be more inclined to use a single cell, maybe exposing two configure methods that allow passing a text or an attributed text.

The SOLID principles are closer to guidelines than rules; they’re impossible to follow 100% of the time, and yet, adherence to 90% of the principles will ultimately serve to create a far superior application. In that way, it’s okay to break the Single Responsibility Principle in this way, and I do without feeling guilty. This is one of these cases, in my opinion, where readability and less code duplication has a better implication than SRP.

mindgraffiti replied:
>In that way, it’s okay to break the Single Responsibility Principle in this way, and I do without feeling guilty

🙂 Thanks Paolo. I'll change it back to using the original cell and delete the code for this one.

mindgraffiti committed:
Make cell configuration functions private

mindgraffiti committed:
Delete commented out code and add TODOs

mindgraffiti committed:
Use the ServiceLocator instead of directly using the FeatureFlag class

mindgraffiti committed:
Add an order with refunds to `orders-load-all.json`

Unit tests will fail

mindgraffiti committed:
Update existing order store unit tests so they pass

mindgraffiti committed:
Merge branch 'develop' into issue/1099-read-only-refunds

mindgraffiti commented:
Thanks @jaclync!

**Net Amount cell:** I did not communicate that very well. The cell is not fully configured. In the testing section, it should simply appear. (I will edit the PR summary to correct this).

**OrderDetailsDataSource refactor:** I will write a P2 post about it and ask for better solutions. I won't merge this PR until that has been resolved.

**OrderStoreTests:** Thanks for the reminder! I will add more objects to the load-all-orders JSON and add some new tests.

kyleaparker reviewed (changes requested):
Looking good just a couple of things: 
- On Android we changed the parentheses around More details to a dash instead. So it would read: Oct 9, 2019 via Credit Card (Stripe) - View details
- Can we remove the line of text under the **Net** section

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource+Sections.swift:
```diff
@@ -0,0 +1,234 @@
+        let payment: Section = {
+            var rows: [Row] = [.payment, .customerPaid]
+
+            if ServiceLocator.featureFlagService.isFeatureFlagEnabled(. refunds) {
```
you might also want to implement the `refunds` feature flag case in `DefaultFeatureFlagService`. otherwise, the default is always on (released to all - we can probably update this to be off by default) 😬

mindgraffiti replied:
🤔That's not good. There are two ways to check if the feature flag is enabled. One is by using the calculated property `.enabled` in FeatureFlag's enum definition. The other is by using the method in the DefaultFeatureFlag service. Do you mind if I removed the calculated property? There are no callers to it.

jaclync commented:
hey @mindgraffiti I responded to the comments, and just 2 things mentioned in the comments earlier:
- Feature flag implementation
- `TwoColumnHeadlineFooterTableViewCell`

> OrderDetailsDataSource refactor: I will write a P2 post about it and ask for better solutions. I won't merge this PR until that has been resolved.

@pmusolino what do you think of the refactoring in this PR, breaking the bigger `OrderDetailsDataSource` class into extensions with some properties exposed? If I were to refactor it, I'll probably have a separate class/struct to generate the sections and another class/struct or cell extension to configure the cells. Do you have other refactoring ideas?

mindgraffiti commented:
Hey @pmusolino @jleandroperez 👋 

I took a naïve approach to breaking down the `OrderDetailsDataSource` in this branch. It was [too large on develop](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource.swift). The downside to what I did is that it exposes several properties and methods to be public.

Do you have any ideas on how to break it apart, while maintaining the correct level of access control?

mindgraffiti committed:
Add refunds to the correct feature flag file

mindgraffiti committed:
Revert "The whole point of creating a second cell was to remove "

This reverts commit e59fb576dc14be7df0a0d26f2a4bf4031e7927f7.

mindgraffiti committed:
The whole point of creating a second cell was to remove 

the extra responsibility that would be found in TwoColumnHeadlineFootnoteTableViewCell.

mindgraffiti committed:
Change the first stackview: remove 6 points spacing

Use standard spacing instead, so that when a label collapses, the stackview collapsing behavior is correct.

mindgraffiti committed:
Remove the footnote text on the Net amount cell

mindgraffiti committed:
Change the first stackview: remove 6 points spacing

in the TwoColumnHeadlineAttributedFooter cell

mindgraffiti committed:
Change the footer text to read " - View details" to match Android change

mindgraffiti committed:
Merge branch 'develop' into issue/1099-read-only-refunds

pmusolino commented:
You do a really great job to split it, it was massive.
Unfortunately, extensions can only reach across to a private declaration if the extension and the declaration live in the same source file.
So I think that we can make private only config functions, as suggested by @jaclync. 
I'm trying to think of another way to split the data source in a better way, and I also do some searches about that, but nothing comes out 🤔 Probably in your place, I would have done the same way.

mindgraffiti committed:
Continue updates to Orders unit tests

mindgraffiti committed:
Update last of the Orders unit tests

jleandroperez commented:
Hola @mindgraffiti !!!! I agree with @pmusolino, I think you did an amazing job at splitting `OrderDetailsDataSource`.

- No matter what... there'll always be ~18 views that will show up in the Order Details
- Those 18 types of views (TableViewCells!) need to be setup
- You've got the intermediate structures that let you simplify things up (Sections / Rows)

Few random ideas (that **definitely** don't need to be addressed, just thinking out loud!):

- `OrderDetailsDataSource+Sections`: That right there is an adapter! And could be unit tested, which could give you peace of mind when introducing new changes.

 **Input:** An Order instance. **Output:** Sections + Rows. That could potentially live in its own class:

```
class OrderDetailsAdapter {
   struct Section { }
   enum Row { }

   func sections(from order: Order) -> [Section] { }
}
```

(OR something like that, just a sketch!).

- `checkIfCopyingIsAllowed`: That could be a `Row` property itself


---

**But no matter what** the number of lines of code won't shrink. All you can do is shuffle things around, and make it a bit more solid.

Don't feel overwhelmed by a big file, as long as it's clear what's going on, and can be easily understood. The only thing that could truly make all the things solid, is unit test the expected Sections / Rows, based on an Order. All the rest is just cosmetic.

Great job Thuy!!!

mindgraffiti committed:
Merge branch 'develop' into issue/1099-read-only-refunds

mindgraffiti commented:
@kyleaparker made the changes:
- update text link to `- View details` 
- remove the footer text on the Net amount

<img src="https://user-images.githubusercontent.com/1062444/69279466-084f9300-0baa-11ea-9405-935bb0715b5e.png" width="350" />

(The correct currency symbol for refunds and net amount will arrive in a separate PR. The correct Net amount number will arrive in a separate PR).

mindgraffiti committed:
Code review changes: delete the extra table view cell and modify the old cell

mindgraffiti committed:
Restore OrderDetailsDataSource file so it can be refactored in a separate PR

mindgraffiti commented:
@jaclync adding the refactoring to the code I already had split out (creating an Adapter class and a separate extension / class for configuring the cells) added another thousand or so lines of code that changed. Give this one a look and then I'll merge this PR and refactor the rest of the OrderDetailsDataSource in a separate PR.

mindgraffiti committed:
Merge branch 'develop' into issue/1099-read-only-refunds

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderPaymentDetailsViewModel.swift:
```diff
@@ -82,12 +83,55 @@ final class OrderPaymentDetailsViewModel {
+
+         // Last, style the link text.
+        let linkAttributes: [NSAttributedString.Key: Any] = [
+            .foregroundColor: StyleManager.wooCommerceBrandColor
```
let's use the new color in `UIColor+SemanticColors` from Cesar's Dark Mode color palette work! p99K0U-21F-p2

mindgraffiti replied:
thanks for the reminder +1

jaclync commented on WooCommerce/Classes/ViewRelated/ReusableViews/TwoColumnHeadlineFootnoteTableViewCell.swift:
```diff
@@ -32,15 +34,16 @@ final class TwoColumnHeadlineFootnoteTableViewCell: UITableViewCell {
+    }
+
+    func toggleFootnote() {
+        footnoteLabel.isHidden = !footnoteLabel.isHidden
```
hm I was wondering when this would be the use case? since this cell class `TwoColumnHeadlineFootnoteTableViewCell` could be reused for different index paths, the `footnoteLabel` could be shown/hidden by other index paths before

mindgraffiti replied:
changed to `hideFootnote()`

jaclync commented on WooCommerce/Classes/ViewModels/OrderDetailsViewModel/OrderDetailsDataSource+Constants.swift:
```diff
@@ -0,0 +1,140 @@
+
+
+// MARK: - Constants
+extension OrderDetailsDataSource {
```
could these constants also go back to the main file and back to private?

mindgraffiti replied:
yep! I thought we had other files where the Constants were kept separately, but searching the project shows we don't.

jaclync reviewed (commented):
it's so easy to review the changes for refunds now! thanks for taking the time to move the code back, just few questions and LGTM after!

- I was a bit confused about the use case for `TwoColumnHeadlineFootnoteTableViewCell `'s `toggleFootnote`
- could the Constants go back to `OrderDetailsDataSource` main file for this PR?
- let's use the new semantic color in `UIColor+SemanticColors` from Cesar's Dark Mode work

mindgraffiti committed:
Merge branch 'develop' into issue/1099-read-only-refunds

mindgraffiti committed:
Change method name to hideFootnote()

mindgraffiti committed:
Use semantic colors

mindgraffiti committed:
Remove Constants file, add constants to data source

It's not possible to make the constants extension private, but it is inside the same file as the rest of the code, which makes access internal-only.
